lib/kube/proxy/sess.go

```diff
@@ -0,0 +1,1193 @@
+func (s *session) allParticipants() []string {
+	var participants []string
+	for _, p := range s.partiesHistorical {
+		participants = append(participants, p.Ctx.User.GetName())
+	}
+
+	return participants
+}
+
+// canStart checks if a session can start with the current set of participants.
+func (s *session) canStart() (bool, auth.PolicyOptions, error) {
+	var participants []auth.SessionAccessContext
+	for _, party := range s.parties {
+		if party.Ctx.User.GetName() == s.ctx.User.GetName() {
+			continue
+		}
+
+		roleNames := party.Ctx.Identity.GetIdentity().Groups
+		roles, err := getRolesByName(s.forwarder, roleNames)
+		if err != nil {
+			return false, auth.PolicyOptions{}, trace.Wrap(err)
+		}
+
+		participants = append(participants, auth.SessionAccessContext{Roles: roles})
+	}
+
+	yes, options, err := s.accessEvaluator.FulfilledFor(participants)
+	return yes, options, trace.Wrap(err)
+}
+
+// Close terminates a session and disconnects all participants.
+func (s *session) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.closeOnce.Do(func() {
+		s.io.BroadcastMessage("Closing session...")
+		s.stateUpdate.L.Lock()
+		defer s.stateUpdate.L.Unlock()
+		s.state = types.SessionState_SessionStateTerminated
+		s.io.Close()
+		s.stateUpdate.Broadcast()
+		err := s.trackerUpdateState(types.SessionState_SessionStateTerminated)
+		if err != nil {
+			s.log.Warnf("Failed to set tracker state to %v", types.SessionState_SessionStateTerminated)
+		}
+
+		s.log.Debugf("Closing session %v.", s.id.String())
+		close(s.closeC)
+		for id, party := range s.parties {
+			err = party.Close()
+			if err != nil {
+				s.log.WithError(err).Errorf("Failed to disconnect party %v", id.String())
+			}
+		}
+
+		if s.recorder != nil {
+			s.recorder.Close(s.forwarder.ctx)
+		}
+	})
+
+	return nil
+}
+
+func getRolesByName(forwarder *Forwarder, roleNames []string) ([]types.Role, error) {
+	var roles []types.Role
+
+	for _, roleName := range roleNames {
+		role, err := forwarder.cfg.CachingAuthClient.GetRole(context.TODO(), roleName)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		roles = append(roles, role)
+	}
+
+	return roles, nil
+}
+
+func (s *session) trackerGet() (types.SessionTracker, error) {
+	sess, err := s.forwarder.cfg.AuthClient.GetSessionTracker(s.forwarder.ctx, s.id.String())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return sess, nil
+}
+
+func (s *session) trackerCreate(p *party, policySets []*types.SessionTrackerPolicySet) error {
+	initator := &types.Participant{
+		ID:         p.ID.String(),
+		User:       p.Ctx.User.GetName(),
+		LastActive: time.Now().UTC(),
+	}
+
+	req := &proto.CreateSessionTrackerRequest{
+		ID:                s.id.String(),
+		Namespace:         defaults.Namespace,
+		Type:              string(types.KubernetesSessionKind),
+		Hostname:          s.podName,
+		ClusterName:       s.ctx.teleportCluster.name,
+		Login:             "root",
```

Comment:
Hard coded "root" intentional here?

Reply:
Yes. Kubeadm translates this to a `docker exec` call with the user being root as default for all I know. That and kubectl doesn't provide a way to specify the user. I am not aware of a case of `kubectl exec` not resulting in a root shell.

Reply:
It was but perhaps it would be better leaving this empty. The user you get in as is up to the pod and kubeadm just translates this to a docker exec without the user argument. The kubectl CLI nor API provides a way to set the login either.

Reply:
Ok, sounds good to me.